### PR TITLE
feat(balancer) dynamic upstream keepalive pools

### LINKED ISSUE #2835: Hostname Provided via SNI and hostname provided via HTTP are different
### Summary

After setting up kong to serve multiple APIs, i am getting this error when i try to call API 2 just after a call to API 1. 
I saw in the logs that the hostname of the first API is used while requesting the second API. 

I tried the patch [#2129 ](https://github.com/Mashape/kong/issues/2129) but still getting the same error

### Steps To Reproduce

1. Setup Kong
2. Add apis using the admin interface : 
       curl -i -X POST localhost:8001/apis  --data 'name=vcmp1' --data 'upstream_url=https://lb-vcmp01.**'*  --data 'hosts=lb-vcmp01.***'
       curl -i -X POST localhost:8001/apis --data 'name=vcmp2' --data 'upstream_url=https://lb-vcmp02.***'  --data 'hosts=lb-vcmp02.***'

3. Call the API 1 then API 2 : 
    - curl -I https://localhost:8000 --header 'Host: lb-vcmp01.***' 
    - curl -I https://localhost:8000 --header 'Host: lb-vcmp02.***'

### Additional Details & Logs

- Kong version (0.10.3)
- Using official docker image
- I should also mention that all the APIs are hosted in the same machine

### LINKED ISSUE #3788: Keepalive to Kubernetes Edge Routes causing Incorrect Routing
### Summary
Our Kong has several separate services that route to a RedHat kubernetes re-skin called Openshift Origin (OSO).

The way OSO works, is that edge routes are defined for the separate services hosted on it - that all resolve to the same IP (OSO Router). The way the router determines which pod to send traffic to, is via the HTTP HOST Header.

So:
```
curl https://myapp.oso.company.com - routes to myapp pod
curl https://theirapp.oso.company.com - routes to theirapp pod
```

But
```
nslookup myapp.oso.company.com = 10.0.0.1
nslookup theirapp.oso.company.com = 10.0.0.1
```

so if I wanted to use the IP to make the call, I’d do this:
```
curl https://10.0.0.1/ -H "Host: myapp.oso.company.com"
curl https://10.0.0.1/ -H "Host: theirapp.oso.company.com"
```

The problem is this: **When the proxy for myapp.oso has >5tps, Kong will sometimes send traffic for the theirapp proxy to the myapp.oso upstream**. When we disable ngx upstream keepalives, the issue is resolved.

My theory, is that because Keepalives are maintained as a IP:PORT pair, and we are sending to the same IP:PORT for both myapp and theirapp - Kong is re-using the socket for myapp which was for theirapp's traffic.

There is nothing in the logs, even at the most detailed - when this occurs.

### Steps To Reproduce

1. Create two services on Kubernetes with SSL edge routes and create proxies for them on Kong
2. Send >5tps traffic to one of them
3. Test calls against the other, from the same kong cluster

### Additional Details & Logs

- Kong version: 0.14.1
- Kong debug-level startup logs (`$ kong start --vv`) - DEBUG LOGS DO NOT SHOW ANYTHING
- Kong error logs (`<KONG_PREFIX>/logs/error.log`) - ERROR LOGS DO NOT SHOW ANYTHING
- Kong configuration (the output of a GET request to Kong's Admin 
```json
{
  "plugins": {
    "enabled_in_cluster": [
      "acl",
      "request-size-limiting",
      "oauth2",
      "kong-path-based-routing",
      "cors",
      "kong-oidc-auth",
      "rate-limiting",
      "kong-upstream-jwt",
      "kong-oidc-multi-idp",
      "kong-spec-expose",
      "jwt",
      "request-termination",
      "kong-cluster-drain",
      "statsd",
      "correlation-id",
      "kong-splunk-log"
    ],
    "available_on_server": {
      "kong-path-based-routing": true,
      "kong-spec-expose": true,
      "kong-cluster-drain": true,
      "correlation-id": true,
      "kong-splunk-log": true,
      "jwt": true,
      "cors": true,
      "rate-limiting": true,
      "kong-oidc-auth": true,
      "kong-upstream-jwt": true,
      "request-size-limiting": true,
      "request-termination": true,
      "kong-oidc-multi-idp": true,
      "kong-service-virtualization": true,
      "request-transformer": true,
      "acl": true,
      "statsd": true,
      "oauth2": true
    }
  },
  "tagline": "Welcome to kong",
  "configuration": {
    "plugins": [
      "request-transformer",
      "kong-service-virtualization",
      "kong-cluster-drain",
      "kong-upstream-jwt",
      "kong-splunk-log",
      "kong-spec-expose",
      "kong-oidc-auth",
      "kong-path-based-routing",
      "kong-oidc-multi-idp",
      "correlation-id",
      "oauth2",
      "statsd",
      "jwt",
      "rate-limiting",
      "acl",
      "request-size-limiting",
      "request-termination",
      "cors"
    ],
    "admin_ssl_enabled": false,
    "lua_ssl_verify_depth": 3,
    "trusted_ips": {
      
    },
    "lua_ssl_trusted_certificate": "\/usr\/local\/kong\/ssl\/kongcert.pem",
    "loaded_plugins": {
      "kong-path-based-routing": true,
      "kong-spec-expose": true,
      "kong-cluster-drain": true,
      "correlation-id": true,
      "kong-splunk-log": true,
      "jwt": true,
      "cors": true,
      "kong-oidc-multi-idp": true,
      "kong-oidc-auth": true,
      "kong-upstream-jwt": true,
      "acl": true,
      "oauth2": true,
      "statsd": true,
      "kong-service-virtualization": true,
      "request-transformer": true,
      "request-size-limiting": true,
      "rate-limiting": true,
      "request-termination": true
    },
    "cassandra_username": "xxx",
    "admin_ssl_cert_csr_default": "\/usr\/local\/kong\/ssl\/admin-kong-default.csr",
    "ssl_cert_key": "\/usr\/local\/kong\/ssl\/kongprivatekey.key",
    "dns_resolver": {
      
    },
    "pg_user": "kong",
    "mem_cache_size": "1024m",
    "cassandra_data_centers": [
      "dc1:2",
      "dc2:3"
    ],
    "nginx_admin_directives": {
      
    },
    "cassandra_password": "******",
    "custom_plugins": {
      
    },
    "pg_host": "127.0.0.1",
    "nginx_acc_logs": "\/usr\/local\/kong\/logs\/access.log",
    "proxy_listen": [
      "0.0.0.0:8000",
      "0.0.0.0:8443 ssl http2"
    ],
    "client_ssl_cert_default": "\/usr\/local\/kong\/ssl\/kong-default.crt",
    "ssl_cert_key_default": "\/usr\/local\/kong\/ssl\/kong-default.key",
    "dns_no_sync": false,
    "db_update_propagation": 5,
    "nginx_err_logs": "\/usr\/local\/kong\/logs\/error.log",
    "cassandra_port": 9042,
    "dns_order": [
      "LAST",
      "SRV",
      "A",
      "CNAME"
    ],
    "dns_error_ttl": 1,
    "headers": [
      "latency_tokens"
    ],
    "dns_stale_ttl": 4,
    "nginx_optimizations": true,
    "database": "cassandra",
    "pg_database": "kong",
    "nginx_worker_processes": "auto",
    "lua_package_cpath": "",
    "admin_acc_logs": "\/usr\/local\/kong\/logs\/admin_access.log",
    "lua_package_path": ".\/?.lua;.\/?\/init.lua;",
    "nginx_pid": "\/usr\/local\/kong\/pids\/nginx.pid",
    "upstream_keepalive": 120,
    "cassandra_contact_points": [
      "xxx",
      "xxx",
      "xxx",
      "xxx",
      "xxx",
      "xxx"
    ],
    "admin_access_log": "off",
    "client_ssl_cert_csr_default": "\/usr\/local\/kong\/ssl\/kong-default.csr",
    "proxy_listeners": [
      {
        "ssl": false,
        "ip": "0.0.0.0",
        "proxy_protocol": false,
        "port": 8000,
        "http2": false,
        "listener": "0.0.0.0:8000"
      },
      {
        "ssl": true,
        "ip": "0.0.0.0",
        "proxy_protocol": false,
        "port": 8443,
        "http2": true,
        "listener": "0.0.0.0:8443 ssl http2"
      }
    ],
    "proxy_ssl_enabled": true,
    "proxy_access_log": "off",
    "ssl_cert_csr_default": "\/usr\/local\/kong\/ssl\/kong-default.csr",
    "enabled_headers": {
      "latency_tokens": true,
      "X-Upstream-Status": false,
      "X-Proxy-Latency": true,
      "server_tokens": false,
      "Server": false,
      "Via": false,
      "X-Upstream-Latency": true
    },
    "cassandra_ssl": true,
    "cassandra_local_datacenter": "ELR",
    "db_resurrect_ttl": 30,
    "db_update_frequency": 5,
    "cassandra_consistency": "LOCAL_QUORUM",
    "client_max_body_size": "1000m",
    "admin_error_log": "\/dev\/stderr",
    "pg_ssl_verify": false,
    "dns_not_found_ttl": 30,
    "pg_ssl": false,
    "cassandra_repl_factor": 1,
    "cassandra_lb_policy": "RequestDCAwareRoundRobin",
    "cassandra_repl_strategy": "SimpleStrategy",
    "nginx_kong_conf": "\/usr\/local\/kong\/nginx-kong.conf",
    "error_default_type": "text\/plain",
    "nginx_http_directives": {
      
    },
    "real_ip_header": "X-Real-IP",
    "kong_env": "\/usr\/local\/kong\/.kong_env",
    "cassandra_schema_consensus_timeout": 10000,
    "dns_hostsfile": "\/etc\/hosts",
    "admin_listeners": [
      {
        "ssl": false,
        "ip": "0.0.0.0",
        "proxy_protocol": false,
        "port": 8001,
        "http2": false,
        "listener": "0.0.0.0:8001"
      }
    ],
    "ssl_ciphers": "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256",
    "ssl_cert": "\/usr\/local\/kong\/ssl\/kongcert.crt",
    "prefix": "\/usr\/local\/kong",
    "admin_ssl_cert_key_default": "\/usr\/local\/kong\/ssl\/admin-kong-default.key",
    "cassandra_ssl_verify": true,
    "db_cache_ttl": 0,
    "ssl_cipher_suite": "modern",
    "real_ip_recursive": "off",
    "proxy_error_log": "\/dev\/stderr",
    "client_ssl_cert_key_default": "\/usr\/local\/kong\/ssl\/kong-default.key",
    "nginx_daemon": "off",
    "anonymous_reports": false,
    "cassandra_timeout": 5000,
    "nginx_proxy_directives": {
      
    },
    "pg_port": 5432,
    "log_level": "notice",
    "client_body_buffer_size": "20m",
    "client_ssl": false,
    "lua_socket_pool_size": 30,
    "admin_ssl_cert_default": "\/usr\/local\/kong\/ssl\/admin-kong-default.crt",
    "cassandra_keyspace": "kong_stage",
    "ssl_cert_default": "\/usr\/local\/kong\/ssl\/kong-default.crt",
    "nginx_conf": "\/usr\/local\/kong\/nginx.conf",
    "admin_listen": [
      "0.0.0.0:8001"
    ]
  },
  "version": "0.14.1",
  "node_id": "3a7824e4-2971-4daa-9351-c05293b26f04",
  "lua_version": "LuaJIT 2.1.0-beta3",
  "prng_seeds": {
    "pid: 30": 121842322351,
    "pid: 37": 205914716415,
    "pid: 36": 619910823160,
    "pid: 35": 158818822910,
    "pid: 33": 106581771972,
    "pid: 34": 119182521233,
    "pid: 31": 230127232253,
    "pid: 32": 194211052318
  },
  "timers": {
    "pending": 4,
    "running": 0
  },
  "hostname": "kong-186-txzsv"
}
```
- Operating system: Alpine Linux kong-186-txzsv 3.10.0-693.el7.x86_64

### LINKED ISSUE #4001: Hostname Provided via SNI and hostname provided via HTTP are different (updated)
### Summary

This issue is similar/same as #2835. It seems that Kong fails sometimes to set SNI hostname properly.

### Steps To Reproduce

1. Setup Kong (I've used 0.14.1 on CentOS 7).

2. Add two APIs with upstream set towards Apache via https, ideally on same IP.

`curl -X POST http://localhost:8001/apis -d name=partners -d uris=/partners -d upstream_url=https://partners-backend.loc/partners`

`curl -X POST http://localhost:8001/apis -d name=feedback -d uris=/feedback -d upstream_url=https://feedback-backend.loc/tl/feedback`

3. Call first API, then immediately after second API.

`curl http://localhost:8000/partners/foo`
`curl http://localhost:8000/feedback/bar`

First query will succeed, second fail will error "400 Bad Request".

Apache will log related error message:

`AH02032: Hostname partners-backend.loc provided via SNI and hostname feedback-backend.loc provided via HTTP are different`

which I understand as the Kong gateway send request towards `feedback-backend.loc` Host with `partners-backend.loc` SNI and that was rejected.

Please note that if you add certain delay between queries (e.g. about 1 minute), the second query succeeds.

Also if you repeat the second query, it will succeed as well.

### Additional Details & Logs

I've tried to fix the error by myself, but ended up in the middle of debugging.

The error is closely related to 

`proxy_set_header   Connection        $upstream_connection;`

set it `nginx_kong.lua`; by forcing the Connection header to 'close' (or removing it to use the same NGINX default value, 'close'), the setup behaves correctly (so this is the functional workaround).

By adding NGINX access logs:

`log_format upstreamlog '[$time_local] $remote_addr to: $upstream_addr: $request request Connection header (http_connection) $http_connection, upstream request Connection header (upstream_http_connection) $upstream_http_connection, response Connection header (sent_http_connection) $sent_http_connection';`

`access_log nginx-debug-access.log upstreamlog;`

`[20/Nov/2018:11:41:32 -0500] 127.0.0.1 to: 10.60.48.201:443: GET /partners/foo HTTP/1.1 request Connection header (http_connection) -, upstream request Connection header (upstream_http_connection) Keep-Alive, response Connection header (sent_http_connection) close
[20/Nov/2018:11:41:35 -0500] 127.0.0.1 to: 10.60.48.201:443: GET /feedback/foo HTTP/1.1 request Connection header (http_connection) -, upstream request Connection header (upstream_http_connection) close, response Connection header (sent_http_connection) close`

above, second query failed, apache logged SNI error

now, kong restarted, delay between queries:

`[20/Nov/2018:11:41:51 -0500] 127.0.0.1 to: 10.60.48.201:443: GET /partners/foo HTTP/1.1 request Connection header (http_connection) -, upstream request Connection header (upstream_http_connection) Keep-Alive, response Connection header (sent_http_connection) close
[20/Nov/2018:11:42:45 -0500] 127.0.0.1 to: 10.60.48.201:443: GET /feedback/foo HTTP/1.1 request Connection header (http_connection) -, upstream request Connection header (upstream_http_connection) Keep-Alive, response Connection header (sent_http_connection) close`

second query succeeded, no SNI error

My understanding is that the `$upstream_connection` gets filled by `Kong.access()`:

        access_by_lua_block {
            Kong.access()
        }

so I've added a debug print into runloop/handler.lua to very the code is really setting only Keep-Alive:

     `-- Keep-Alive and WebSocket Protocol Upgrade Headers
      if var.http_upgrade and lower(var.http_upgrade) == "websocket" then
        var.upstream_connection = "upgrade"
        var.upstream_upgrade    = "websocket"

      else
        var.upstream_connection = "keep-alive"
      end
      print ("DEBUG: upstream_connection variable set to " .. var.upstream_connection)
      [...]
      print ("DEBUG: safe check, upstream_connection variable set to " .. var.upstream_connection)
    end,

and it seems to work well (variable always set to keep-alive):

`2018/11/20 11:41:31 [notice] 24582#0: *15 [lua] handler.lua:604: before(): DEBUG: upstream_connection variable set to keep-alive, client: 127.0.0.1, server: kong, request: "GET /partners/foo HTTP/1.1", host: "localhost:8000"
2018/11/20 11:41:31 [notice] 24582#0: *15 [lua] handler.lua:620: before(): DEBUG: safe check, upstream_connection variable set to keep-alive, client: 127.0.0.1, server: kong, request: "GET /partners/foo HTTP/1.1", host: "localhost:8000"`

`2018/11/20 11:41:35 [notice] 24582#0: *28 [lua] handler.lua:604: before(): DEBUG: upstream_connection variable set to keep-alive, client: 127.0.0.1, server: kong, request: "GET /feedback/foo HTTP/1.1", host: "localhost:8000"
2018/11/20 11:41:35 [notice] 24582#0: *28 [lua] handler.lua:620: before(): DEBUG: safe check, upstream_connection variable set to keep-alive, client: 127.0.0.1, server: kong, request: "GET /feedback/foo HTTP/1.1", host: "localhost:8000"`

while the NGINX log shows "close" was sent in the second query (repeated line from above):

`[20/Nov/2018:11:41:35 -0500] 127.0.0.1 to: 10.60.48.201:443: GET /feedback/foo HTTP/1.1 request Connection header (http_connection) -, upstream request Connection header (upstream_http_connection) close, response Connection header (sent_http_connection) close`

I've tried to increase log level of NGINX to debug, but that doesn't show how exactly is the variable var.upstream_connection passed to NGINX.

(EDIT: slipped enter & formatting)

### PR DESCRIPTION
**Pull request by thibaultcha**
This replaces the keepalive implementation from ngx_http_upstream_module with the one proposed by openresty/lua-nginx-module#1600. It is proposed against `next` due to the potentially disruptive nature of such a large change in production environments. It is my advice that this change be first introduced in a release candidate.

For the time being, the ngx_http_lua and lua-resty-core patches have been applied to our openresty-build-tools repository and are applied when building OpenResty for Kong. The final API may be subject to changes by OpenResty, but incorporating the changes in Kong ahead of time reduces possible virtual host confusion issues and/or attack vectors.

Sibling PR with the aforementioned patches: https://github.com/Kong/kong-build-tools/pull/257

The proposed OpenResty patches offer a way to circumvent the NGINX limitations detailed in the following tickets:

* https://trac.nginx.org/nginx/ticket/1340
* https://trac.nginx.org/nginx/ticket/1593

The behavior changes in Kong from this work are:

* Avoid virtual host confusion when proxying upstream virtual services over TLS and mTLS (based on SNI and/or client certificate id).
* Allow for an indefinite idle timeout and max requests for connections in the keepalive pool (a behavior reverted by NGINX in 1.15.3).
* Introduce new properties for upstream keepalive pooling.
* Ensure older properties are properly deprecated and that their settings cascades to the new ones.
* Use different examples for NGINX directives injection.

In the future, we could implement per-Service (or per-Upstream) keepalive pools, allowing users to granularly enable/disable, or configure upstream keepalive pools as they wish.

Fix #2835
Fix #3788
Fix #4001

cc @jeremyjpj0916 @rsbrisci

### CODE REVIEW

File: kong/conf_loader.lua [hunk 1]
```diff
@@ -231,53 +292,66 @@ local CONF_INFERENCES = {
...
-          conf.nginx_upstream_keepalive_timeout = tostring(conf.nginx_http_upstream_keepalive_timeout)
-        end
-      end,
+      replacement = "upstream_keepalive_idle_timeout",
+      alias = upstream_keepalive_deprecated_properties,
     }
   },
-  nginx_http_upstream_keepalive_requests = { -- TODO: remove since deprecated in 2.0
+
+  -- TODO: remove since deprecated in X.Y
```
Comment (thibaultcha) [inline kong/conf_loader.lua:328]: Pending on the release including this patch.

**Comment by jeremyjpj0916** [comment]
Just gonna say, this is awesome. Thanks for putting this together!

File: kong/init.lua [hunk 2]
```diff
@@ -811,13 +818,32 @@ function Kong.balancer()
...
+  local kong_conf = kong.configuration
+
+  if enable_keepalive and kong_conf.upstream_keepalive_pool_size > 0
+     and subsystem == "http"
+  then
+    local pool = balancer_data.ip .. "|" .. balancer_data.port
+
+    if balancer_data.scheme == "https" then
+      -- upstream_host is SNI
+      pool = pool .. "|" .. ngx.var.upstream_host
```
Comment (dndx) [inline kong/init.lua:831]: Should we add the upstream client certificate ID somewhere in here as well?

Comment (thibaultcha) [reply kong/init.lua:831]: Totally agreed, forgot we support upstream mTLS now :+1:

Comment (thibaultcha) [reply kong/init.lua:831]: Done, thanks for the suggestion!

File: kong/conf_loader/init.lua [hunk 3]
```diff
@@ -233,53 +294,66 @@ local CONF_INFERENCES = {
...
-          conf.nginx_upstream_keepalive_timeout = tostring(conf.nginx_http_upstream_keepalive_timeout)
-        end
-      end,
+      replacement = "upstream_keepalive_idle_timeout",
+      alias = upstream_keepalive_deprecated_properties,
     }
   },
-  nginx_http_upstream_keepalive_requests = { -- TODO: remove since deprecated in 2.0
+
+  -- TODO: remove since deprecated in X.Y
```
Comment (thibaultcha) [inline kong/conf_loader/init.lua:330]: I need to force-push a version of the commit in which this says 2.1

**Pull request merged**